### PR TITLE
Implement options for 'noise' operator to insert final punctuation.

### PR DESCRIPTION
**Pull request by natsegal**

### CODE REVIEW

File: nmtwizard/preprocess/operators/noise.py [hunk 1]
```diff
@@ -206,6 +226,22 @@ def _apply_word_noise(self, tokens):
...
+            and random.random() <= self._final_punct_add_prob
+            and all(
+                punct not in src_tokens[-1].surface for punct in self._final_punct_list
+            )
+            and all(
+                punct not in tgt_tokens[-1].surface for punct in self._final_punct_list
+            )
+        ):
+            punct_token = pyonmttok.Token(random.choice(self._final_punct_list))
+            punct_token.join_left = True
```
Comment (guillaumekln) [inline nmtwizard/preprocess/operators/noise.py:241]: This is not true for all punctuations and languages. For example !?:; in French should have a leading space.

The user should be able to configure a leading space for one side only (source or target). Maybe you can accept tuples in the punctuation list to configure different source and target values, e.g. `[".", (";", " ;"), (":", " :")]`. If there is a leading space, you should remove it and not set `join_left` when creating the new token.

What do you think?

File: nmtwizard/preprocess/operators/noise.py [hunk 2]
```diff
@@ -92,9 +94,14 @@ def __init__(self, config, *args):
         self._char_equivalence_prob = config.get("char_equivalence_prob", 0)
         self._char_equivalence_table = config.get("char_equivalence_table", {})
         self._add_marker = config.get("add_marker", 0)
+        self._final_punct_add_prob = config.get("final_punct_add_prob", 0)
+        if self._final_punct_add_prob:
+            self._final_punct_list = config.get("final_punct_list") or [".", ";", ":"]
```
Comment (guillaumekln) [inline nmtwizard/preprocess/operators/noise.py:99]: I think there should not be a default value based on the other comment.

File: nmtwizard/preprocess/operators/noise.py [hunk 3]
```diff
@@ -206,6 +226,22 @@ def _apply_word_noise(self, tokens):
                 new_tokens.append(token)
         return new_tokens
 
+    def _apply_bilingual_noise(self, src_tokens, tgt_tokens):
+        if (
+            self._final_punct_add_prob > 0
+            and random.random() <= self._final_punct_add_prob
+            and all(
```
Comment (guillaumekln) [inline nmtwizard/preprocess/operators/noise.py:233]: To be safe, also check that `src_tokens` and `tgt_tokens` are not empty.

**Comment by natsegal** [comment]
What do you think now ?

**Review by guillaumekln — APPROVED** [review]

**Pull request merged**